**Name:** Stav Yosef

**ID:** 316298876


---


**Name:** Daniel Sabba

**ID:** 311500227


**Colab link:** [https://drive.google.com/file/d/1PxyFTxrHWmrt9IFB9XtJr-jM5yk9owVM/view?usp=sharing](https://drive.google.com/file/d/1PxyFTxrHWmrt9IFB9XtJr-jM5yk9owVM/view?usp=sharing) 

<h2> Content Based Recommender System - Metafeatures </h2>

The goal of this notebook is to implement content based recommender system on the Movielens 100k dataset.

The movie profile is based on the movie genres.

Two approaches are implemented. 

<b> Approach 1: </b>

The user profile is either a weighted average of the movie profile he\she rated, or the average of the movie profile he\she liked (rating >=3) - the average rating he\she didn't like (with a lower weight for the disliked movies)

The recommended movies are the closest ones (e.g. by Cosine similarity) to the user profile vector

The implementation is based on this blog post [website]
    
<b> Approach 2: </b>

The similarity score between two movies is calculated by computing the similarity between the movie profiles of each movies pair. 

The predicted rating a user will give to a candidate item, is calculated by the rating the user gave to K most similar items to the candidate item. The recommended movies are those with highest predicted rating.  

The implementation is based on this post [website2]

[website2]: https://www.kaggle.com/varian97/item-based-collaborative-filtering    

[website]: https://towardsdatascience.com/movie-recommendation-system-based-on-movielens-ef0df580cd0e

In [1]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.metrics.pairwise import cosine_similarity

# Setup

In [106]:
import os.path as path
import warnings

import numpy as np
import pandas as pd
from sklearn.metrics.pairwise import cosine_similarity

warnings.filterwarnings('ignore')

## Download Dataset

In [107]:
!wget http://files.grouplens.org/datasets/movielens/ml-100k.zip

--2021-01-10 13:28:32--  http://files.grouplens.org/datasets/movielens/ml-100k.zip
Resolving files.grouplens.org (files.grouplens.org)... 128.101.65.152
Connecting to files.grouplens.org (files.grouplens.org)|128.101.65.152|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 4924029 (4.7M) [application/zip]
Saving to: ‘ml-100k.zip.2’

ml-100k.zip.2       100%[===================>]   4.70M  14.0MB/s    in 0.3s    

2021-01-10 13:28:32 (14.0 MB/s) - ‘ml-100k.zip.2’ saved [4924029/4924029]



In [108]:
!ls

item_vec_latent.npy  ml-100k.zip    ml-100k.zip.2
ml-100k		     ml-100k.zip.1  sample_data


In [109]:
!unzip -u ml-100k.zip

Archive:  ml-100k.zip


## Data Managment

In [110]:
!ls "ml-100k/"

allbut.pl  u1.base  u2.test  u4.base  u5.test  ub.base	u.genre  u.occupation
mku.sh	   u1.test  u3.base  u4.test  ua.base  ub.test	u.info	 u.user
README	   u2.base  u3.test  u5.base  ua.test  u.data	u.item


In [111]:
def get_dataset_folder() -> str:
    return "ml-100k/"

In [112]:
def get_train_test_path(k: int) -> (str, str):
    if 1 <= k <= 5:
        return path.join(get_dataset_folder(), f'u{k}.base'), path.join(get_dataset_folder(), f'u{k}.test')
    else:
        return None

In [113]:
def build_matrix(_path: str):
    df = pd.read_csv(_path, sep='\t', names=['user_id', 'item_id', 'rating', 'timestamp'])

    n_users = int(get_num_users())
    n_items = int(get_num_items())

    matrix = np.zeros((n_users, n_items))
    for row in df.itertuples():
        matrix[row[1] - 1, row[2] - 1] = row[3]
    return matrix

In [114]:
def build_train_test_matrix(k: int) -> (np.ndarray, np.ndarray):
    path_train, path_test = get_train_test_path(k=k)

    return build_matrix(_path=path_train), build_matrix(_path=path_test)

In [115]:
def build_train_matrix() -> np.ndarray:
    matrix_train, _ = build_train_test_matrix(1)

    for i in range(2, 6, 1):
        train, _ = build_train_test_matrix(i)
        matrix_train += train
    
    return matrix_train

In [116]:
def build_test_matrix() -> np.ndarray:
    _, matrix_test = build_train_test_matrix(1)

    for i in range(2, 6, 1):
        _, test = build_train_test_matrix(i)
        matrix_test += test
    
    return matrix_test

In [117]:
def get_data_path() -> str:
    return path.join(get_dataset_folder(), "u.data")

In [118]:
def get_users_path() -> str:
    return path.join(get_dataset_folder(), "u.user")

In [119]:
def get_genres_path() -> str:
    return path.join(get_dataset_folder(), "u.genre")

In [120]:
def get_items_path() -> str:
    return path.join(get_dataset_folder(), "u.item")

In [121]:
def load_genres() -> list:
    _ = pd.read_csv(get_genres_path(),
                    delimiter='|',
                    names=["Genre", "Code"],
                    encoding='latin-1')
    
    return _[_.columns[0]].to_list()

In [122]:
def load_items() -> pd.DataFrame:
    m_cols = ['movie_id', 'movie_title', 'release date', 'video release date', 'IMDb URL'] + load_genres()
    return pd.read_csv(get_items_path(), delimiter='|', names=m_cols, encoding='latin-1')

In [123]:
def load_data() -> pd.DataFrame:
    m_cols = ['user_id', 'item_id', 'rating', 'timestamp']
    return pd.read_csv(get_data_path(), delimiter='\t', names=m_cols, encoding='latin-1')

In [124]:
def load_users() -> pd.DataFrame:
    m_cols = ['user_id', 'age', 'gender', 'occupation', 'zip_code']
    return pd.read_csv(get_users_path(), delimiter='\t', names=m_cols, encoding='latin-1')

In [125]:
def get_num_users() -> int:
    users = load_users()
    return users['user_id'].unique().shape[0]

In [126]:
def get_num_items() -> int:
    items = load_items()
    return items['movie_id'].unique().shape[0]

# Data loading

In [127]:
column_names = ['user_id', 'item_id', 'rating', 'timestamp']
ratings = load_data()

ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype
---  ------     --------------   -----
 0   user_id    100000 non-null  int64
 1   item_id    100000 non-null  int64
 2   rating     100000 non-null  int64
 3   timestamp  100000 non-null  int64
dtypes: int64(4)
memory usage: 3.1 MB


In [128]:
#Binary option should be set to True if the rating should be binary. 
#It should be set to True for the first approach and False for the second approach
BINARY_OPTION = True
#NEGATIVE WEIGHT is relevant only for the first approach
NEGATIVE_WEIGHT = 0.25

In [129]:
def brating(row):
    if row['rating'] >= 3:
        val = 1
    elif row['rating'] >=0:
        val = -NEGATIVE_WEIGHT
    else:
        val = row['rating']
    return val


ratings['binary_rating'] = ratings.apply(brating, axis=1)
ratings.head()

user_id  item_id  rating  timestamp  binary_rating
0      196      242       3  881250949           1.00
1      186      302       3  891717742           1.00
2       22      377       1  878887116          -0.25
3      244       51       2  880606923          -0.25
4      166      346       1  886397596          -0.25

In [130]:
movie_titles = load_items()
movie_titles

movie_id                                movie_title  ... War  Western
0            1                           Toy Story (1995)  ...   0        0
1            2                           GoldenEye (1995)  ...   0        0
2            3                          Four Rooms (1995)  ...   0        0
3            4                          Get Shorty (1995)  ...   0        0
4            5                             Copycat (1995)  ...   0        0
...        ...                                        ...  ...  ..      ...
1677      1678                          Mat' i syn (1997)  ...   0        0
1678      1679                           B. Monkey (1998)  ...   0        0
1679      1680                       Sliding Doors (1998)  ...   0        0
1680      1681                        You So Crazy (1994)  ...   0        0
1681      1682  Scream of Stone (Schrei aus Stein) (1991)  ...   0        0

[1682 rows x 24 columns]

# Question 1

## A

Creating movie profile, removing all columns but the genres.

In [131]:
movie_profile = movie_titles[load_genres()[1:]]
movie_profile.sort_index(axis=0, inplace=True)
movie_profile.head()

Action  Adventure  Animation  Children's  ...  Sci-Fi  Thriller  War  Western
0       0          0          1           1  ...       0         0    0        0
1       1          1          0           0  ...       0         1    0        0
2       0          0          0           0  ...       0         1    0        0
3       1          0          0           0  ...       0         0    0        0
4       0          0          0           0  ...       0         1    0        0

[5 rows x 18 columns]

## B

In [132]:
def get_movies_table_from_ids(df: pd.DataFrame) -> pd.DataFrame:
    movies_trim = movie_titles[movie_titles.columns.drop(load_genres())]
    
    df = df.merge(movies_trim, how='inner', on='movie_id')

    return df

In [133]:
def get_movies_cosine_similarity() -> pd.DataFrame:
    _ = movie_profile
    return pd.DataFrame(cosine_similarity(_, _),
                        index=_.index + 1,
                        columns=_.index + 1)

In [134]:
def find_most_similar_movies(movie_id: int, top_k: int) -> pd.Series:
    df_cos = get_movies_cosine_similarity()

    movie_vec = df_cos[movie_id]  # Extract the movie similarity cosine vector
    movie_vec = movie_vec.drop(movie_id)  # Remove the movie itself from the vector

    top_movies = movie_vec.sort_values(ascending=False).iloc[:top_k]  # Get top top_k similar movies.

    res = pd.DataFrame({"movie_id": top_movies.index, "similartiy_score": top_movies})
    res.reset_index(drop=True, inplace=True)

    return get_movies_table_from_ids(df=res)

In [155]:
top_similar = find_most_similar_movies(movie_id=1, top_k=5)
top_similar.head(5)

movie_id  ...                                           IMDb URL
0       422  ...  http://us.imdb.com/M/title-exact?Aladdin%20and...
1      1219  ...  http://us.imdb.com/M/title-exact?Goofy%20Movie...
2        95  ...  http://us.imdb.com/M/title-exact?Aladdin%20(1992)
3      1078  ...  http://us.imdb.com/M/title-exact?Oliver%20&%20...
4       477  ...  http://us.imdb.com/M/title-exact?Matilda%20(1996)

[5 rows x 6 columns]

## C

We understood this question in two different ways so down below we solved it twice just in case.

### First approach

Use function b to find top k (5) items similar for specific item.

In [158]:
top_similar = find_most_similar_movies(movie_id=1, top_k=5)
top_similar.head()

movie_id  ...                                           IMDb URL
0       422  ...  http://us.imdb.com/M/title-exact?Aladdin%20and...
1      1219  ...  http://us.imdb.com/M/title-exact?Goofy%20Movie...
2        95  ...  http://us.imdb.com/M/title-exact?Aladdin%20(1992)
3      1078  ...  http://us.imdb.com/M/title-exact?Oliver%20&%20...
4       477  ...  http://us.imdb.com/M/title-exact?Matilda%20(1996)

[5 rows x 6 columns]

In [173]:
top_similar = find_most_similar_movies(movie_id=1000, top_k=5)
top_similar.head()

movie_id  ...                                           IMDb URL
0       575  ...  http://us.imdb.com/M/title-exact?City%20Slicke...
1       415  ...  http://us.imdb.com/M/title-exact?Apple%20Dumpl...
2      1188  ...  http://us.imdb.com/M/title-exact?Young%20Guns%...
3       435  ...  http://us.imdb.com/M/title-exact?Butch%20Cassi...
4        73  ...  http://us.imdb.com/M/title-exact?Maverick%20(1...

[5 rows x 6 columns]

#### Explanation

In [137]:
movie_titles[movie_titles["movie_id"] == 1]

movie_id       movie_title release date  ...  Thriller War  Western
0         1  Toy Story (1995)  01-Jan-1995  ...         0   0        0

[1 rows x 24 columns]

As we can see the genres of Toy Story are:

1.   Animation
2.   Children's
3.   Comedy

In [138]:
movie_titles[movie_titles["movie_id"].isin([422, 1219, 95, 1078, 477])]

movie_id                             movie_title  ... War  Western
94          95                          Aladdin (1992)  ...   0        0
421        422  Aladdin and the King of Thieves (1996)  ...   0        0
476        477                          Matilda (1996)  ...   0        0
1077      1078                 Oliver & Company (1988)  ...   0        0
1218      1219                   Goofy Movie, A (1995)  ...   0        0

[5 rows x 24 columns]

Aladdin and the King of Thieves have the same exactly genres as Toy Story so the similarity is 1.

Goofy Movie, A & Aladdin have the same genres as Toy Story with 1 more genre so it's close to 1 but not one (0.866)

Oliver & Company & Matilda have 2/3 genres as Toy Story so they are pretty similar but not exactly the same $\rightarrow$ 0.816


**Good Results!**



---



In [174]:
movie_titles[movie_titles["movie_id"] == 1000]

movie_id            movie_title release date  ...  Thriller War  Western
999      1000  Lightning Jack (1994)  01-Jan-1994  ...         0   0        1

[1 rows x 24 columns]

As we can see the genres of Lightning Jack are:

1.   Comedy
2.   Western

In [175]:
movie_titles[movie_titles["movie_id"].isin([575, 415, 1188, 435, 73])]

movie_id                                        movie_title  ... War  Western
72          73                                    Maverick (1994)  ...   0        1
414        415                    Apple Dumpling Gang, The (1975)  ...   0        1
434        435          Butch Cassidy and the Sundance Kid (1969)  ...   0        1
574        575  City Slickers II: The Legend of Curly's Gold (...  ...   0        1
1187      1188                               Young Guns II (1990)  ...   0        1

[5 rows x 24 columns]

City Slickers II: The Legend of Curly's Gold have the same exactly genres as Lightning Jack so the similarity is 1.

All others movies are Comedy & Western with addition one genre $\rightarrow$ 0.816


**Good Results!**

### Second approach

We look at the question in different perspective, mabye the question is to take movie X and movie Y and we need to produce common recommendation for both of the movie in the same time.

The idea is to check similarity to each of the items and then check the intersection of them by the movie id column.

In the last stage calculate the average $\frac{score_{i,first-mov-id} + score_{i,second-mov-id}}{2}$ And take the top k (5).

In [139]:
def most_similar_for_2_movies(first_mov_id: int, second_mov_id: int, top_k: int = 5) -> pd.DataFrame:
    df = movie_profile.copy()
    k = df.shape[0]

    similar_movies_first = find_most_similar_movies(movie_id=first_mov_id, top_k=k)
    similar_movies_second = find_most_similar_movies(movie_id=second_mov_id, top_k=k)

    merged = similar_movies_first.merge(similar_movies_second, how='inner', on='movie_id')

    pt = merged.pivot_table(index='movie_id')
    pt = pt.mean(axis=1)
    pt.sort_values(ascending=False, inplace=True)

    res = pt.to_frame(name="score")
    res = res.iloc[:top_k]
    res = pd.DataFrame({"movie_id": res.index, "score": res["score"]})
    res.reset_index(drop=True, inplace=True)
    
    res = get_movies_table_from_ids(res)

    return res

In [140]:
movie_titles[movie_titles['movie_id'] == 1]

movie_id       movie_title release date  ...  Thriller War  Western
0         1  Toy Story (1995)  01-Jan-1995  ...         0   0        0

[1 rows x 24 columns]

In [141]:
movie_titles[movie_titles['movie_id'] == 402]

movie_id   movie_title release date  ...  Thriller War  Western
401       402  Ghost (1990)  01-Jan-1990  ...         1   0        0

[1 rows x 24 columns]

Toy Story genres:

1.   Animation
2.   Children's
3.   Comedy

Ghost genres:

1.   Comedy
2.   Romance
3.   Thriller

In the results we should see something in between Toy Story & Ghost $\rightarrow$ movies with mixed genres.

In [142]:
top_similar = most_similar_for_2_movies(first_mov_id=1, second_mov_id=402, top_k=5)
top_similar.head()

movie_id  ...                                           IMDb URL
0      1219  ...  http://us.imdb.com/M/title-exact?Goofy%20Movie...
1       422  ...  http://us.imdb.com/M/title-exact?Aladdin%20and...
2       490  ...  http://us.imdb.com/M/title-exact?To%20Catch%20...
3       408  ...  http://us.imdb.com/M/title-exact?Close%20Shave...
4        90  ...  http://us.imdb.com/M/title-exact?So%20I%20Marr...

[5 rows x 6 columns]

In [143]:
movie_titles[movie_titles["movie_id"].isin([1219, 422, 490, 408, 90])]

movie_id                             movie_title  ... War  Western
89          90     So I Married an Axe Murderer (1993)  ...   0        0
407        408                   Close Shave, A (1995)  ...   0        0
421        422  Aladdin and the King of Thieves (1996)  ...   0        0
489        490                 To Catch a Thief (1955)  ...   0        0
1218      1219                   Goofy Movie, A (1995)  ...   0        0

[5 rows x 24 columns]

As we can see it's produced movies as we anticipated (mixed genres)

**Good Results!**

## D

In [144]:
with open('item_vec_latent.npy', 'rb') as f:
    items_vec_latent = np.load(f)

items_vec_latent

array([[-0.04783021,  0.02221448, -0.05127662, ..., -0.01007532,
        -0.02563317, -0.01639253],
       [ 0.01588862,  0.02223237,  0.01566593, ...,  0.01292368,
         0.01028191, -0.00914688],
       [ 0.02308113, -0.000877  ,  0.01033436, ...,  0.01291555,
        -0.0148144 , -0.03359524],
       ...,
       [ 0.01516091,  0.04004909,  0.03174036, ...,  0.00368512,
        -0.00010248, -0.04094235],
       [ 0.02120049,  0.01216243, -0.00912177, ..., -0.02031322,
        -0.03148889,  0.01064072],
       [-0.00871174,  0.03308845, -0.00286855, ...,  0.02723331,
         0.00094329,  0.05252173]])

In [145]:
df_items_ex1 = pd.DataFrame(items_vec_latent)
df_items_ex1

0         1         2   ...        17        18        19
0    -0.047830  0.022214 -0.051277  ... -0.010075 -0.025633 -0.016393
1     0.015889  0.022232  0.015666  ...  0.012924  0.010282 -0.009147
2     0.023081 -0.000877  0.010334  ...  0.012916 -0.014814 -0.033595
3    -0.014282  0.004960  0.015986  ...  0.015146 -0.005361  0.012977
4    -0.009577 -0.024733  0.021396  ... -0.010347  0.008591  0.026025
...        ...       ...       ...  ...       ...       ...       ...
1677  0.023184  0.026013  0.014345  ...  0.002501 -0.048414  0.019632
1678 -0.056877  0.037772  0.010527  ... -0.027863  0.033905  0.014389
1679  0.015161  0.040049  0.031740  ...  0.003685 -0.000102 -0.040942
1680  0.021200  0.012162 -0.009122  ... -0.020313 -0.031489  0.010641
1681 -0.008712  0.033088 -0.002869  ...  0.027233  0.000943  0.052522

[1682 rows x 20 columns]

In [146]:
_ = df_items_ex1
df_items_ex1_cos_sim = pd.DataFrame(cosine_similarity(_, _), index=_.index + 1, columns=_.index + 1)
df_items_ex1_cos_sim

1         2         3     ...      1680      1681      1682
1     1.000000 -0.368499 -0.060298  ...  0.240328  0.153660  0.212588
2    -0.368499  1.000000  0.313474  ...  0.102602 -0.029763  0.009031
3    -0.060298  0.313474  1.000000  ...  0.048277  0.141143 -0.140425
4     0.045094  0.033400 -0.184132  ...  0.024685  0.059517 -0.036755
5    -0.148955 -0.368053 -0.281953  ... -0.101527 -0.163029 -0.102740
...        ...       ...       ...  ...       ...       ...       ...
1678  0.094924  0.260929 -0.039231  ...  0.415177  0.488688  0.139502
1679  0.006712  0.197061 -0.462575  ... -0.027497 -0.118878  0.254880
1680  0.240328  0.102602  0.048277  ...  1.000000  0.135019  0.059673
1681  0.153660 -0.029763  0.141143  ...  0.135019  1.000000  0.226509
1682  0.212588  0.009031 -0.140425  ...  0.059673  0.226509  1.000000

[1682 rows x 1682 columns]

In [147]:
def find_most_similar_items_mf(mov_id: int, top_k: int = 5) -> pd.DataFrame:
    item_vec = df_items_ex1_cos_sim[mov_id]  # Extract the movie similarity cosine vector
    item_vec = item_vec.drop(mov_id)  # Remove the movie itself from the vector

    top_items = item_vec.sort_values(ascending=False).iloc[:5]

    res = pd.DataFrame({"movie_id": top_items.index, "similartiy_score": top_items})
    res.reset_index(drop=True, inplace=True)

    return get_movies_table_from_ids(df=res)

In [148]:
find_most_similar_items_mf(mov_id=1, top_k=5)

movie_id  ...                                           IMDb URL
0      1453  ...  http://us.imdb.com/M/title-exact?Angel%20on%20...
1       405  ...  http://us.imdb.com/M/title-exact?Mission:%20Im...
2       511  ...  http://us.imdb.com/M/title-exact?Lawrence%20of...
3       979  ...  http://us.imdb.com/M/title-exact?Trigger%20Eff...
4      1139  ...  http://us.imdb.com/M/title-exact?Hackers%20(1995)

[5 rows x 6 columns]

In [149]:
find_most_similar_items_mf(mov_id=402, top_k=5)

movie_id  ...                                           IMDb URL
0       236  ...  http://us.imdb.com/M/title-exact?Citizen%20Rut...
1      1180  ...  http://us.imdb.com/M/title-exact?I%20Love%20Tr...
2       865  ...  http://us.imdb.com/M/title-exact?Ice+Storm%2C+...
3       591  ...  http://us.imdb.com/M/title-exact?Primal%20Fear...
4       449  ...  http://us.imdb.com/M/title-exact?Star%20Trek:%...

[5 rows x 6 columns]

Implemented also the second approach from C

In [150]:
def most_similar_for_2_items_mf(first_mov_id: int, second_mov_id: int, top_k: int = 5) -> pd.DataFrame:
    df = movie_profile.copy()
    k = df.shape[0]

    similar_items_first = find_most_similar_items_mf(mov_id=first_mov_id, top_k=k)
    similar_items_second = find_most_similar_items_mf(mov_id=second_mov_id, top_k=k)

    merged = similar_items_first.merge(similar_items_second, how='inner', on='movie_id')

    pt = merged.pivot_table(index='movie_id')
    pt = pt.mean(axis=1)
    pt.sort_values(ascending=False, inplace=True)

    res = pt.to_frame(name="score")
    res = res.iloc[:top_k]
    res = pd.DataFrame({"movie_id": res.index, "score": res["score"]})
    res.reset_index(drop=True, inplace=True)

    return get_movies_table_from_ids(res)

In [151]:
most_similar_for_2_items_mf(first_mov_id=1, second_mov_id=402, top_k=5)

Empty DataFrame
Columns: [score, movie_id, movie_title, release date, video release date, IMDb URL]
Index: []

### Explanation

It is obvious that the results are not the same. To understand that we need to understand how the evaluation of each method is different from each other.

In the matrix factorization we calculating each item score (latent vector) based on users rating so two different movies with the same genres can be evaluated differently because one can be highly score rated movie and the other low score rated, On the other hand in content based approach we calculating each item score based on it's "dry" properties, for example genres, title, year of release, producer and more. 

# Question 2

In [44]:
user_x_movie = pd.pivot_table(ratings, values='binary_rating', index=['item_id'], columns = ['user_id'])
user_x_movie.sort_index(axis=0, inplace=True)

userIDs = user_x_movie.columns
user_profile = pd.DataFrame(columns = movie_profile.columns)

user_x_movie is the rating matrix. Rows are item_id, columns are user_id. Missing values are NaN

In [45]:
user_x_movie

user_id  1    2    3    4    5    6    7    ...  937  938  939   940  941  942  943
item_id                                     ...                                    
1        1.0  1.0  NaN  NaN  1.0  1.0  NaN  ...  NaN  1.0  NaN   NaN  1.0  NaN  NaN
2        1.0  NaN  NaN  NaN  1.0  NaN  NaN  ...  NaN  NaN  NaN   NaN  NaN  NaN  1.0
3        1.0  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   NaN  NaN  NaN  NaN
4        1.0  NaN  NaN  NaN  NaN  NaN  1.0  ...  NaN  NaN  NaN -0.25  NaN  NaN  NaN
5        1.0  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   NaN  NaN  NaN  NaN
...      ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   ...  ...  ...  ...
1678     NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   NaN  NaN  NaN  NaN
1679     NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   NaN  NaN  NaN  NaN
1680     NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   NaN  NaN  NaN  NaN
1681     NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   NaN  NaN  NaN  NaN
1682     NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   NaN  NaN  NaN  NaN

[1682 rows x 943 columns]

In [46]:
user_profile.head()

Empty DataFrame
Columns: [Action, Adventure, Animation, Children's, Comedy, Crime, Documentary, Drama, Fantasy, Film-Noir, Horror, Musical, Mystery, Romance, Sci-Fi, Thriller, War, Western]
Index: []

The user profile is the average rating the user gave to movies which belong to each genre

In [47]:
for i in range(len(user_x_movie.columns)):
  working_df = movie_profile.mul(user_x_movie.iloc[:,i], axis=0)
  user_profile.loc[userIDs[i]] = working_df.mean(axis=0)

In [48]:
user_profile.head()

Action  Adventure  Animation  ...  Thriller       War   Western
1  0.219669   0.140625   0.026654  ...  0.135110  0.073529  0.017463
2  0.157258   0.080645   0.016129  ...  0.197581  0.032258  0.016129
3  0.078704   0.032407   0.000000  ...  0.166667 -0.009259  0.000000
4  0.250000   0.125000   0.000000  ...  0.406250  0.083333  0.041667
5  0.158571   0.082857   0.071429  ...  0.078571  0.042857  0.021429

[5 rows x 18 columns]

**TFIDF**

In the movie profile we want to give higher weight to rare genres. The movie profile is now represented by a TFIDF of the genres in the dataset

In [49]:
# TFIDF
df = movie_profile.sum()
idf = (len(movie_titles) / df).apply(np.log) #log inverse of DF
TFIDF = movie_profile.mul(idf.values)

In [50]:
TFIDF.head()

Action  Adventure  Animation  Children's  ...  Sci-Fi  Thriller  War  Western
0  0.000000   0.000000   3.690069    2.623718  ...     0.0  0.000000  0.0      0.0
1  1.902286   2.522464   0.000000    0.000000  ...     0.0  1.902286  0.0      0.0
2  0.000000   0.000000   0.000000    0.000000  ...     0.0  1.902286  0.0      0.0
3  1.902286   0.000000   0.000000    0.000000  ...     0.0  0.000000  0.0      0.0
4  0.000000   0.000000   0.000000    0.000000  ...     0.0  1.902286  0.0      0.0

[5 rows x 18 columns]

The recommended items to a user, are the items with highest Cosine similarity with the user profile vector

In [51]:
cosine_similarity_user_item = cosine_similarity(user_profile, TFIDF)

## A

In [52]:
def predict_most_similar_items_per_user(user_id: int,num_items: int = 5) -> np.ndarray:
    result = np.argsort(cosine_similarity_user_item[user_profile.index.get_loc(user_id),:])[::-1][:num_items]
    ret_result = [movie_profile.index[i] for i in result]
    return np.array(ret_result)

## B

### Test 1


In [53]:
user_idx = 0

In [54]:
s = user_profile.iloc[user_idx]
s.sort_values(ascending=False)

Drama          0.292279
Comedy         0.261949
Action         0.219669
Adventure      0.140625
Thriller       0.135110
Sci-Fi         0.135110
Romance        0.129596
Crime          0.077206
War            0.073529
Children's     0.051471
Horror         0.043199
Animation      0.026654
Musical        0.024816
Western        0.017463
Documentary    0.013787
Mystery        0.013787
Film-Noir      0.003676
Fantasy        0.002757
Name: 1, dtype: float64

We can see that user 1 is into Drama (0.292279), Comedy (0.261949) and Action (0.219669)

Let's predict and examine the results.

In [55]:
res = predict_most_similar_items_per_user(user_id=user_idx + 1, num_items=5)
df_res = pd.DataFrame({"movie_id": res + 1})
df_res = df_res.merge(movie_titles, how='inner', on='movie_id')
df_res

movie_id                          movie_title  ... War  Western
0         4                    Get Shorty (1995)  ...   0        0
1        74  Faster Pussycat! Kill! Kill! (1965)  ...   0        0
2        65   What's Eating Gilbert Grape (1993)  ...   0        0
3       964          Month by the Lake, A (1995)  ...   0        0
4       723              Boys on the Side (1995)  ...   0        0

[5 rows x 24 columns]

As we can see the first 2 movies are Action, Comedy & Drama.

3-5 movies are Comedy & Drama.


**Results are good!**

TFIDF is the movies profile matrix after TFIDF weighting 

In [56]:
TFIDF.iloc[res]

Action  Adventure  Animation  Children's  ...  Sci-Fi  Thriller  War  Western
3    1.902286        0.0        0.0         0.0  ...     0.0       0.0  0.0      0.0
73   1.902286        0.0        0.0         0.0  ...     0.0       0.0  0.0      0.0
64   0.000000        0.0        0.0         0.0  ...     0.0       0.0  0.0      0.0
963  0.000000        0.0        0.0         0.0  ...     0.0       0.0  0.0      0.0
722  0.000000        0.0        0.0         0.0  ...     0.0       0.0  0.0      0.0

[5 rows x 18 columns]

**Good Matches!**

### Test 2


In [57]:
user_idx = 111

In [58]:
s = user_profile.iloc[user_idx]
s.sort_values(ascending=False)

Drama          0.375000
Thriller       0.315217
Action         0.206522
Comedy         0.168478
Romance        0.146739
Children's     0.130435
Mystery        0.125000
Crime          0.125000
Sci-Fi         0.086957
Adventure      0.086957
Fantasy        0.065217
Film-Noir      0.043478
Horror         0.021739
War            0.016304
Documentary    0.000000
Animation      0.000000
Musical        0.000000
Western        0.000000
Name: 112, dtype: float64

We can see that user 112 is into Drama (0.375), Thriller (0.315217) and Action (0.206522)

Let's predict and examine the results.

In [59]:
res = predict_most_similar_items_per_user(user_id=user_idx + 1, num_items=5)
df_res = pd.DataFrame({"movie_id": res + 1})
df_res = df_res.merge(movie_titles, how='inner', on='movie_id')
df_res

movie_id                    movie_title release date  ...  Thriller War  Western
0       337       House of Yes, The (1997)  01-Jan-1997  ...         1   0        0
1        54                Outbreak (1995)  01-Jan-1995  ...         1   0        0
2      1025         Fire Down Below (1997)  01-Jan-1997  ...         1   0        0
3      1491        Tough and Deadly (1995)  01-Jan-1995  ...         1   0        0
4       244  Smilla's Sense of Snow (1997)  14-Mar-1997  ...         1   0        0

[5 rows x 24 columns]

As we can see all of them are Thriller & Drama.

All of them also Comedy or Action which is the number 3 in the preferences of the user


**Results are good!**

TFIDF is the movies profile matrix after TFIDF weighting 

In [60]:
TFIDF.iloc[res]

Action  Adventure  Animation  ...  Thriller  War  Western
336   0.000000        0.0        0.0  ...  1.902286  0.0      0.0
53    1.902286        0.0        0.0  ...  1.902286  0.0      0.0
1024  1.902286        0.0        0.0  ...  1.902286  0.0      0.0
1490  1.902286        0.0        0.0  ...  1.902286  0.0      0.0
243   1.902286        0.0        0.0  ...  1.902286  0.0      0.0

[5 rows x 18 columns]

**Good Matches**

## C


In [61]:
movie_sim_df = pd.DataFrame(cosine_similarity(movie_profile, movie_profile),
                            index=movie_profile.index,
                            columns=movie_profile.index)
movie_sim_df.head()

0         1        2         3     ...      1678      1679     1680     1681
0  1.000000  0.000000  0.00000  0.333333  ...  0.000000  0.000000  0.57735  0.00000
1  0.000000  1.000000  0.57735  0.333333  ...  0.408248  0.000000  0.00000  0.00000
2  0.000000  0.577350  1.00000  0.000000  ...  0.707107  0.000000  0.00000  0.00000
3  0.333333  0.333333  0.00000  1.000000  ...  0.000000  0.408248  0.57735  0.57735
4  0.000000  0.333333  0.57735  0.333333  ...  0.408248  0.408248  0.00000  0.57735

[5 rows x 1682 columns]

In [62]:
r_cols = ['user_id', 'movie_id', 'rating', 'timestamp']
data_path = path.join(get_dataset_folder(), 'u1.base')

f1_train = pd.read_csv(data_path, delimiter='\t', names=r_cols, encoding='latin-1')
f1_train.drop(columns=["timestamp"], inplace=True)
f1_train['binary_rating'] = ratings.apply(brating, axis=1)

user_x_movie = pd.pivot_table(f1_train, values='binary_rating', index=['movie_id'], columns=['user_id'])
user_x_movie_n = user_x_movie.copy()
user_x_movie_n.fillna(0, inplace=True)
user_x_movie_n.head()

user_id    1    2    3    4    5    6    7    ...  937  938  939  940  941  942  943
movie_id                                      ...                                   
1         1.00  1.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  1.0  0.0  0.0  1.0  0.0  0.0
2         1.00  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  1.0
3        -0.25  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0
4        -0.25  0.0  0.0  0.0  0.0  0.0  1.0  ...  0.0  0.0  0.0  1.0  0.0  0.0  0.0
5        -0.25  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0

[5 rows x 943 columns]

In [63]:
def get_topN_rec(df: pd.DataFrame, n: int) -> pd.DataFrame:
    df2 = df
    df2 = df2.fillna(0)
    x = df2.T

    result = pd.DataFrame(np.zeros((0, n)), columns=[n for n in range(1, n + 1)])

    for i in x.columns:
        df1row = pd.DataFrame(x.nlargest(n, i).index.tolist(), index=[n for n in range(1, n + 1)], columns=[i]).T
        result = pd.concat([result, df1row], axis=0)
        
    return result

In [74]:
def calc_mrr_score(topN: pd.DataFrame) -> pd.DataFrame:
    top2N = topN.copy()
    for top in top2N.columns:
        for user in top2N.index:
            try:
                rating = \
                f1_test[(f1_test["user_id"] == user) & (f1_test["movie_id"] == top2N.at[user, top])].rating.values[-1]
            except:
                rating = -1
            if rating > 3:
                top2N.at[user, top] = 1
            elif rating == -1:
                top2N.at[user, top] = -1
            else:
                top2N.at[user, top] = 0

    mrr_calc = []
    for idx, row in top2N.iterrows():
        try:
            first_occ = list(mrr_score.loc[idx]).index(1) + 1
            mrr_calc.append(1 / first_occ)
        except:
            mrr_calc.append(0)

    top2N["MRR Score"] = mrr_calc
    return top2N

In [75]:
def get_similar_movie(movie_id) -> (pd.DataFrame, pd.DataFrame):
    if movie_id not in movie_profile.index:
        print(movie_id, " not in movie_profile")
        return None, None
    else:
        sim_movie = movie_sim_df.sort_values(by=movie_id, ascending=False).index[1:]
        sim_score = movie_sim_df.sort_values(by=movie_id, ascending=False).loc[:, movie_id].tolist()[1:]
        return sim_movie, sim_score

In [76]:
# predict the rating of movie x by user y
def predict_rating(user_id: int, movie_id: int, max_neighbor: int = 10) -> np.ndarray:
    movies, scores = get_similar_movie(movie_id)
    movie_arr = []
    sim_arr = []
    for movie, score in zip(movies, scores):
        if (movie in user_x_movie_n.index):
            movie_arr.append(movie)
            sim_arr.append(score)

    sim_arr = np.array([x for x in scores])
    movie_arr = np.array([x for x in movie_arr])

    # select only the movies that has already rated by user x
    filtering = user_x_movie_n[user_id].loc[movie_arr] > 0
    indxs = filtering[filtering == True].index
    # calculate the predicted score
    s = 0.0
    # don't estimate rating by less than 4 nearest neighbors (by content)
    if np.sum(sim_arr[indxs][:max_neighbor]) > 0.0 and np.where(sim_arr[indxs] > 0.0)[0].size > 3:
        s = np.dot(sim_arr[indxs][:max_neighbor], user_x_movie_n[user_id].loc[movie_arr[indxs][:max_neighbor]]) \
            / np.sum(sim_arr[indxs][:max_neighbor])

    return s

In [77]:
data_path = path.join(get_dataset_folder(), 'u1.test')

f1_test = pd.read_csv(data_path, delimiter='\t', names=r_cols, encoding='latin-1')
f1_test.drop(columns=["timestamp"], inplace=True)

In [78]:
def predict_values(f1_test: pd.DataFrame) -> pd.DataFrame:
    df = pd.DataFrame(index=set(f1_test["user_id"]), columns=set(f1_test["movie_id"]))
    for idx, row in f1_test.iterrows():
        user, item = row["user_id"], row["movie_id"]
        try:
            pred_value = predict_rating(user, item)
        except:
            # print("user {} or item {} are not exists in U or V".format(user,item))
            continue;
        df.at[user, item] = pred_value
    return df

In [69]:
df = predict_values(f1_test=f1_test)

In [79]:
topN = get_topN_rec(df=df, n=5)

In [80]:
mrr_score = calc_mrr_score(topN=topN)

In [81]:
mrr_score = mrr_score[mrr_score.sum(axis = 1) != -5]

In [85]:
print("MRR:", np.round(mrr_score["MRR Score"].mean(), 3))

MRR: 0.701


## D
As we can see above with content base recommender we achieved 0.701 with MRR evaluation and in matrix factorization we recieved 0.90127 score.

Using matrix factorization, recommendations are based on movies each user watched, and therefore, most viewed movies are more potential to be recommended while their score was high, while in collaborative filtering, a variety of movies are recommended, therefore, when calculating the MRR score, a lot of "rare" movies will be recommended, and therefore the chances of a user to score those items is low.


## E

Content based are better in few aspects, content based filtering isn't sensitive to **cold start problem**, because when a new movie is added to the list, he will be recommended based on his properties and not based on ratings.

Second, in content based we can add more information about the movies or the users in order to further enhance our system.

Content based is easier to explain because each recommendation is based on its properties and therefore easily to interpret, on the other hand in MF we have latent vector, which it's operate like a black box.

One of the most biggest advantages of matrix factorization approach is that the accuracy if we have enough data (for example in our case rating) it is much better and accurate. We can achieve pretty high results because the MF look at each user and customizing recommendation more personally than content based but it's cost us much more computing time against content based approach.
